url64b10c0106_schlabberdog_blocks_tgz-pJ8-com_github_users_schlabberdog_blocks_ui_CliUIJ8/src/com/github/users/schlabberdog/blocks/board/Board.java

```diff
@@ -0,0 +1,232 @@
+		    sb.append('/');
+	    }
+
+        return sb.toString().replace('\0',' '); //die NULs sind unveränderte felder im array
+    }
+
+    public synchronized void print(PrintStream ps) {
+        //zuerst eine map für die elemente erzeugen
+        char[][] map = new char[height][width];
+
+        //oberen rand drucken
+        ps.print('+');
+        for (int i = 0; i < width; i++) {
+            ps.print('-');
+        }
+        ps.print('+');
+        ps.println();
+
+        //alle elemente auf die map packen
+	    for (Map.Entry<Block, Coord> e : blockMap.entrySet()) {
+		    e.getKey().printOntoMap(e.getValue(), map);
+	    }
+
+        //reihenweise drucken
+        for (int y = 0; y < height; y++) {
+            ps.print('|');
+            //spalten drucken
+            for (int x = 0; x < width; x++) {
+                char c = map[y][x];
+                if(c == '\0')
+                    c = ' '; //leere flächen mit leerzeichen füllen
+                ps.print(c);
+            }
+            ps.println('|');
+        }
+
+        //unteren rand drucken
+        ps.print('+');
+        for (int i = 0; i < width; i++) {
+            ps.print('-');
+        }
+        ps.print('+');
+        ps.println();
+    }
+
+    /**
+     * Gibt eine Liste aller Möglichen Schritte aus dem aktuellen Boardzustand an
+     * @return .
+     */
+    public synchronized ArrayList<IMove> getAlternatives() {
+        ArrayList<IMove> alts = new ArrayList<IMove>();
+	    for (Map.Entry<Block, Coord> e : blockMap.entrySet()) {
+			e.getKey().addAlts(e.getValue(),this,alts);
+	    }
+        return alts;
+    }
+
+    /**
+     * Erzeugt ein Snapshot des Boards im aktuellen Zustand
+     * @return Der Snapshot
+     */
+    public synchronized BoardSave getSave() {
+        BoardSave save = new BoardSave();
+
+	    for (Map.Entry<Block, Coord> e : blockMap.entrySet()) {
+		    save.put(e.getKey(),e.getValue());
+	    }
+
+        return save;
+    }
+
+    /**
+     * Wendet einen Zug auf das aktuelle Board an
+     * @param alt Der Zug
+     */
+    public void applyMove(IMove alt) {
+        alt.apply(this);
+    }
+
+    public synchronized void clear() {
+        for (Coord coord : blockMap.values()) {
+            blockMapReverse[coord.x][coord.y] = null;
+        }
+
+        blockMap.clear();
+    }
+
+    /**
+     * Stellt einen Snapshot des Boards wieder her
+     * @param save Der Snapshot, der wiederhergestellt werden soll
+     */
+    public synchronized void applySave(BoardSave save) {
+        //alle aktuellen blocks entfernen
+        clear();
+        //alle blocks im save wiederherstellen
+        for (Map.Entry<Block, Coord> e : save.entrySet()) {
+			insertUnchecked(e.getKey(),e.getValue());
+        }
+    }
+
+	public synchronized Coord getBlockCoord(Block block) {
+		return blockMap.get(block);
+	}
+
+	public Board copy() {
+		Board out = new Board(width,height);
+
+		for (Map.Entry<Block, Coord> e : blockMap.entrySet()) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **226**. At this line the following field is not protected by synchronization: `blockMap`. Your code typically protects this field with synchronization on object: `Board.this`.

**Fix**  
 Consider protecting the above field with synchronization on object: `Board.this`. For example, like at lines: **220**, **203**, **56**, **170**, **199**, **54**, **76**, **93**, **183**, and **67**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **35**, **46**, **54**, **56**, **67**, **76**, **93**, **112**, **138**, **170**, **183**, **199**, **203**, and **220**.